gradle-jdks-certs/src/main/resources/gradle-jdk-resolver.sh

```diff
@@ -0,0 +1,128 @@
+#         * functions;
+#         * expansions «$var», «${var}», «${var%suffix}», and «$( cmd )»;
+#         * various built-in commands including «command» and «set».
+#
+##############################################################################
+
+tmp_work_dir=$(mktemp -d)
+
+die () {
+    echo
+    echo "$*"
+    echo
+    # TODO(crogoz): Cleaning up
+    echo rm -rf "$tmp_work_dir"
+    exit 1
+} >&2
+
+# OS specific support
+case "$( uname )" in                          #(
+  Linux* )          os_name="linux"  ;;       #(
+  Darwin* )         os_name="macos"  ;;       #(
+  * )               die "ERROR Unsupported OS: $( uname )" ;;
+esac
+
+if [ "$os_name" = "linux" ]; then
+    if [ -f /etc/alpine-release ]; then
+      os_name="alpine-linux"
+    fi
+fi
+
+# Arch specific support, see: gradle-jdks:com.palantir.gradle.jdks.CurrentArch.java
+case "$(uname -m)" in                         #(
+  x86-64* )       arch_name="x86-64"  ;;      #(
+  x64* )          arch_name="x86-64"  ;;      #(
+  amd64* )        arch_name="x86-64"  ;;      #(
+  arm64* )        arch_name="aarch64"  ;;     #(
+  arm* )          arch_name="aarch64"  ;;     #(
+  aarch64* )      arch_name="aarch64"  ;;     #(
+  x86* )          arch_name="x86"  ;;         #(
+  i686* )         arch_name="x86"  ;;         #(
+  * )             die "ERROR Unsupported architecture: $( uname -m )" ;;
+esac
+
+read -r major_version < gradle/gradle-jdk-major-version.txt
+read -r distribution_url < gradle/jdks/"$major_version"/$os_name/$arch_name/download-url.txt
+read -r distribution_local_path < gradle/jdks/"$major_version"/$os_name/$arch_name/local-path.txt
+
+regex_pattern="^([a-zA-Z0-9_-]+)-([0-9.]+)-([a-z0-9]+)$"
+
+# Check if the input string matches the regex pattern
+if echo "$distribution_local_path" | grep -q -E "$regex_pattern"; then
+  name=$(echo "$distribution_local_path" | sed -n -E "s/$regex_pattern/\1/p")
+  version=$(echo "$distribution_local_path" | sed -n -E "s/$regex_pattern/\2/p")
+  hash=$(echo "$distribution_local_path" | sed -n -E "s/$regex_pattern/\3/p")
+else
+  die "ERROR: $distribution_local_path for gradle-jdk-major-version=$major_version and $(uname -m) is invalid, expected to match pattern: $regex_pattern"
+fi
+
+# Check if distribution exists in $HOME/.gradle/gradle-jdks
+jdk_installation_directory=$HOME/.gradle/gradle-jdks/$name-$version-$hash
+temp_java_tar=$tmp_work_dir/$name-$version.tar.gz
+die "$distribution_url"
+if [ -d "$jdk_installation_directory" ]; then
+  echo "Distribution $distribution_url already exists in $HOME/.gradle/gradle-jdks, setting JAVA_HOME to $jdk_installation_directory"
+else
+  echo "Distribution $distribution_url does not exist, installing in temporary location $temp_java_tar"
+  # Download and extract the distribution into a temporary directory
+  if command -v curl > /dev/null 2>&1; then
+    echo "Using curl to download $distribution_url"
+    curl -k -C - -o "$temp_java_tar" "$distribution_url"
```

Comment:
Rather than using `temp_java_tar` we can probably just pipe directly in `tar -xf`. This way we avoid needing to deal with a temporary file, and it's a bit faster too it can start untarring before it's finished downloading.

It'd be great if we could add `set -o pipefail` to better handle errors when using pipes, but I think this is a `bash` only feature. I'm not 100% sure how to handle the the case where `curl` fails in `curl | tar`, but I think you can do something like `{ curl; curl_exit=$? } | tar` then check the exit code.